src/multiselect_block_dragger.js

```diff
@@ -0,0 +1,194 @@
+            this.workspace_));
+      });
+    } else {
+      // Dragging a new block that not in the selection list would
+      // clear the multiple selections and only drag that block.
+      blockSelection.forEach((id) => {
+        const element = this.workspace_.getBlockById(id);
+        if (element) {
+          element.pathObject.updateSelected(false);
+        }
+      });
+      blockSelection.clear();
+      blockDraggerList.push(new Blockly.BlockDragger(this.block_,
+          this.workspace_));
+      this.block_.pathObject.updateSelected(true);
+    }
+
+    if (blockDraggerList.length > 1) {
+      // Disabled the highlighting around connection for multiple blocks
+      // dragging because of the bugs.
+      Blockly.RenderedConnection.prototype.highlight = function() {};
+    }
+
+    blockDraggerList.forEach((blockDragger) => {
+      blockDragger.startDrag(currentDragDeltaXY, healStack);
+      this.blockDraggers_.add(blockDragger);
+    });
+  }
+
+  /**
+   * Moves the block to the specified location.
+   * @param {!Event} e The mouseup/touchend event.
+   * @param {!Blockly.utils.Coordinate} currentDragDeltaXY How far the pointer
+   *     has moved from the position at the start of the drag, in pixel units.
+   */
+  drag(e, currentDragDeltaXY) {
+    this.blockDraggers_.forEach(function(blockDragger_) {
+      blockDragger_.drag(e, currentDragDeltaXY);
+    });
+    e.preventDefault();
+    e.stopPropagation();
+  }
+
+  /**
+   * Finishes the block drag.
+   * @param {!Event} e The mouseup/touchend event.
+   * @param {!Blockly.utils.Coordinate} currentDragDeltaXY How far the pointer
+   *     has moved from the position at the start of the drag, in pixel units.
+   */
+  endDrag(e, currentDragDeltaXY) {
+    if (this.blockDraggers_.size > 1) {
+      this.patchUpdateBlockAfterMove(true);
+    }
+    this.blockDraggers_.forEach((blockDragger_) => {
+      if (Blockly.Events.getGroup()) {
+        this.group_ = Blockly.Events.getGroup();
+      } else {
+        Blockly.Events.setGroup(this.group_);
+      }
+      blockDragger_.endDrag(e, currentDragDeltaXY);
+    });
+    if (this.blockDraggers_.size > 1) {
+      // Restore the highlighting around connection for multiple blocks.
+      Blockly.RenderedConnection.prototype.highlight = this.origHighlight_;
+      this.patchUpdateBlockAfterMove(false);
+    }
+  }
+
+  /**
+   * Patch the Blockly.BlockDragger.updateBlockAfterMove_ function.
+   * @param {boolean} on To start the patch or restore.
+   */
+  patchUpdateBlockAfterMove(on) {
+    if (!on) {
+      Blockly.BlockDragger.prototype.updateBlockAfterMove_ =
+          this.origUpdateBlockAfterMove_;
+    } else {
+      Blockly.BlockDragger.prototype.updateBlockAfterMove_ = function(delta) {
+        this.draggingBlock_.moveConnections(delta.x, delta.y);
+        this.fireMoveEvent_();
+        if (this.draggedConnectionManager_.wouldConnectBlock()) {
```

Comment:
Even though this ends in an underscore, it's fine to access it because [it's protected](https://github.com/google/blockly/blob/61322294da70c770b32abb28b9d695af75f61a16/core/block_dragger.js#L71).